modules/stochastic_tools/src/surrogates/MultiOutputGaussianProcess.C

```diff
@@ -0,0 +1,102 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "MultiOutputGaussianProcess.h"
+#include "Sampler.h"
+
+#include "CovarianceFunctionBase.h"
+
+registerMooseObject("StochasticToolsApp", MultiOutputGaussianProcess);
+
+InputParameters
+MultiOutputGaussianProcess::validParams()
+{
+  InputParameters params = SurrogateModel::validParams();
+  params.addClassDescription(
+      "Computes and evaluates Multi-Output Gaussian Process surrogate model.");
+  return params;
+}
+
+MultiOutputGaussianProcess::MultiOutputGaussianProcess(const InputParameters & parameters)
+  : SurrogateModel(parameters),
+    CovarianceInterface(parameters),
+    OutputCovarianceInterface(parameters),
+    _mogp_handler(
+        declareModelData<StochasticTools::MultiOutputGaussianProcessHandler>("_mogp_handler")),
+    _training_params(getModelData<RealEigenMatrix>("_training_params")),
+    _kappa_cho_decomp(_mogp_handler.getKappaCholeskyDecomp()),
+    _K_train(_mogp_handler.getK()),
+    _B(_mogp_handler.getB()),
+    _kappa_results_solve(_mogp_handler.getKappaResultsSolve())
+{
+}
+
+void
+MultiOutputGaussianProcess::evaluate(const std::vector<Real> & x, std::vector<Real> & y) const
+{
+  // Gather the required inputs and matrices
+  unsigned int n_params = _training_params.cols();
+  mooseAssert(x.size() == n_params,
+              "Number of parameters provided for evaluation does not match number of parameters "
+              "used for training.");
+
+  // Gather the test points and standardize
+  RealEigenMatrix test_points(1, n_params);
+  for (unsigned int ii = 0; ii < n_params; ++ii)
+    test_points(0, ii) = x[ii];
+  _mogp_handler.getParamStandardizer().getStandardized(test_points);
+
+  // Compute full test-train covariance and arrange it properly
+  RealEigenMatrix K_train_test(test_points.rows(), _K_train.rows());
+  _mogp_handler.getCovarFunction().computeCovarianceMatrix(
+      K_train_test, test_points, _training_params, false);
+  RealEigenMatrix kappa_train_test;
+  _mogp_handler.getOutputCovar().computeFullCovarianceMatrix(kappa_train_test, K_train_test, _B);
+  RealEigenMatrix kappa_train_test_new(kappa_train_test.rows(), kappa_train_test.cols());
+  RealEigenMatrix tmp;
+  for (unsigned int j = 0; j < kappa_train_test.rows(); ++j)
+  {
+    unsigned int count = 0;
+    for (unsigned int i = 0; i < kappa_train_test.rows(); ++i)
+    {
+      tmp = kappa_train_test(j, Eigen::seq(i, kappa_train_test.cols() - 1, _B.rows()));
+      kappa_train_test_new(j, Eigen::seq(i * count, i * count + tmp.cols() - 1)) = tmp;
+      count += tmp.cols();
+    }
+  }
+
```

Comment:
Please add comments to show what is happening with respect to the theory. (For potential users in the future)